controllers/networkfence_controller.go

```diff
@@ -18,38 +18,135 @@ package controllers
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.10.0/pkg/reconcile
 func (r *NetworkFenceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	// fetch NetworkFence object instance
+	nwFence := &csiaddonsv1alpha1.NetworkFence{}
+	err := r.Get(ctx, req.NamespacedName, nwFence)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			logger.Info("NetworkFence resource not found")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// validate NetworkFence object so as its parameters are neither empty nor nil.
+	err = validateNetworkFenceSpec(nwFence)
+	if err != nil {
+		logger.Error(err, "failed to validate NetworkFence spec")
+
+		errMessage := util.GetErrorMessage(err)
+		nwFence.Status.Result = csiaddonsv1alpha1.FencingOperationResultFailed
+		nwFence.Status.Message = fmt.Sprintf("Failed to validate Networkfence parameters: %v", errMessage)
+		statusErr := r.Client.Status().Update(ctx, nwFence)
+		if statusErr != nil {
+			logger.Error(statusErr, "Failed to update status")
+
+			return ctrl.Result{}, statusErr
+		}
+
+		// invalid parameter, do not requeue
+		return ctrl.Result{}, nil
+	}
+
+	logger = logger.WithValues("Driver Name", nwFence.Spec.Driver, "CIDRs", nwFence.Spec.Cidrs)
 
-	// TODO(user): your logic here
+	conns, unlock := r.Connpool.Get()
+	var (
+		controllerClient *grpc.ClientConn
+	)
+	for _, v := range *conns {
+		if v.DriverName == nwFence.Spec.Driver {
+			for _, cap := range v.Capabilities {
+				if cap.GetNetworkFence() == nil {
+					continue
+				}
+				// check if the object is being deleted
+				if nwFence.GetDeletionTimestamp().IsZero() {
+					if err := r.addFinalizerToNetworkFence(ctx, &logger, nwFence); err != nil {
+						logger.Error(err, "Failed to add NetworkFence finalizer")
+						return reconcile.Result{}, err
+					}
+				} else {
+					if util.ContainsInSlice(nwFence.GetFinalizers(), networkFenceFinalizer) {
+						logger.Info("UnfenceClusterNetwork request")
+						err := r.sendUnfenceNetworkReq(ctx, nwFence, v.Client, logger)
+						if err != nil {
+							logger.Error(err, "failed to unfence cluster network")
+							return ctrl.Result{}, err
+						}
+						// once all finalizers have been removed, the object will then be
+						// deleted.
+						logger.Info("Removing finalizer")
+						if err := r.removeFinalizerFromNetworkFence(ctx, &logger, nwFence); err != nil {
+							logger.Error(err, "Failed to remove NetworkFence finalizer")
+							return reconcile.Result{}, err
+						}
+					}
+					logger.Info("NetworkFence object is terminated, skipping reconciliation")
+					return ctrl.Result{}, nil
+				}
+
+				if cap.GetNetworkFence().GetType() == identity.Capability_NetworkFence_NETWORK_FENCE {
```

Comment:
do the inverse check, skip/return when the type is not what you need, and continue normally below the if (no `else` needed).